src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/activities/ActivitiesService.kt

```diff
@@ -32,57 +34,53 @@ class ActivitiesService(
-    activitiesApiService.getActivitySchedule(event.additionalInformation.activityScheduleId).run {
-      val activity = activitiesApiService.getActivity(activity.id)
-
-      val telemetryMap = mutableMapOf(
-        "activityScheduleId" to id.toString(),
-        "description" to description,
+    synchronise {
+      name = "activity"
+      telemetryClient = this@ActivitiesService.telemetryClient
+      retryQueueService = activitiesUpdateQueueService
+      eventTelemetry = mapOf(
+        "activityScheduleId" to event.additionalInformation.activityScheduleId.toString(),
       )
 
-      // to protect against repeated create messages for same activity
-      if (mappingService.getMappingGivenActivityScheduleIdOrNull(id) != null) {
-        log.warn("Mapping already exists for activity schedule id $id")
-        return
+      checkMappingDoesNotExist {
+        mappingService.getMappingGivenActivityScheduleIdOrNull(event.additionalInformation.activityScheduleId)
       }
-
-      val nomisResponse = try {
-        nomisApiService.createActivity(toNomisActivity(this, activity))
-      } catch (e: Exception) {
-        telemetryClient.trackEvent("activity-create-failed", telemetryMap)
-        log.error("createActivity() Unexpected exception", e)
-        throw e
+      transform {
+        activitiesApiService.getActivitySchedule(event.additionalInformation.activityScheduleId).let { activitySchedule ->
+          eventTelemetry += "description" to activitySchedule.description
+
+          createTransformedActivity(eventTelemetry, activitySchedule).let { nomisResponse ->
+            ActivityMappingDto(
+              nomisCourseActivityId = nomisResponse.courseActivityId,
+              activityScheduleId = event.additionalInformation.activityScheduleId,
+              mappingType = "ACTIVITY_CREATED",
+            )
+          }
+        }
       }
+      saveMapping { mappingService.createMapping(it) }
+    }
+  }
 
-      telemetryMap["courseActivityId"] = nomisResponse.courseActivityId.toString()
-
-      val mapping = ActivityMappingDto(
-        nomisCourseActivityId = nomisResponse.courseActivityId,
-        activityScheduleId = id,
-        mappingType = "ACTIVITY_CREATED",
-      )
-      tryCreateMapping(
-        mapping,
-        MappingTelemetry(failureName = "activity-create-map-failed", attributes = telemetryMap),
-      ) {
-        mappingService.createMapping(mapping)
-      }
+  suspend fun createTransformedActivity(telemetryMap: Map<String, String>, activitySchedule: ActivitySchedule): CreateActivityResponse {
+    val activity = activitiesApiService.getActivity(activitySchedule.activity.id)
 
-      telemetryClient.trackEvent("activity-created-event", telemetryMap)
+    return try {
+      nomisApiService.createActivity(toNomisActivity(activitySchedule, activity))
+    } catch (e: Exception) {
+      telemetryClient.trackEvent("activity-create-failed", telemetryMap)
```

Comment:
not convinced we need this telemetry or the try catch since we have the DLQ and we have exceptions - and this just gets retried. I would like to know what others think. I have kept it for now

Reply:
Happy to go with removing - you have used the telemetry in prod so you'd know :)